src/device.rs

```diff
@@ -835,78 +835,107 @@ impl<'a> ProgrammableStageDescriptor {
 }
 
 #[repr(C)]
-pub struct VertexBufferDescriptor {
-    pub stride: wgt::BufferAddress,
+pub struct VertexBufferLayout {
+    pub array_stride: wgt::BufferAddress,
     pub step_mode: wgt::InputStepMode,
-    pub attributes: *const wgt::VertexAttributeDescriptor,
-    pub attributes_length: usize,
+    pub attributes: *const wgt::VertexAttribute,
+    pub attributes_count: usize,
+}
+
+impl VertexBufferLayout {
+    unsafe fn into_wgpu(&self) -> wgc::pipeline::VertexBufferLayout {
+        wgc::pipeline::VertexBufferLayout {
+            array_stride: self.array_stride,
+            step_mode: self.step_mode,
+            attributes: Cow::Borrowed(slice::from_raw_parts(
```

Comment:
Here and everywhere else, we should be careful about getting a slice like this.
In Vulkan or OpenGL, if the count is 0, then the pointer should no be used. It can be garbage. In Rust, however, `slice::from_raw_parts` expects the pointer to not be garbage even if the size is 0.

I suggest avoiding the calls to `from_raw_parts` directly and instead using a helper:
```rust
unsafe fn make_slice<'a, T: 'a>(pointer: *const T, count: usize) -> &'a [T] {
    if count == 0 {
        &[]
    } else {
        slice::from_raw_parts(pointer, count)
    }
}
```

Reply:
Makes sense, should I replace the 12 usages of `from_raw_parts` in device.rs?

Reply:
yes, that would be good!